core/embed/bootloader/messages.c

```diff
@@ -707,6 +709,26 @@ int process_msg_FirmwareUpload(uint8_t iface_num, uint32_t msg_size,
     return UPLOAD_ERR_FIRMWARE_TOO_BIG;
   }
 
+  // check padding between headers and the code
```

Comment:
It seems to me that it would be possible to put this check inside the block above (just after `read_offset = 0`). We could then get rid of the new global variable `headers_end` - it could be local to the block where this check is implemented.

Reply:
Now that I see it, how about moving the check to the beginning of the function, right after `check_image_header_sig()` and before parsing `current_vhdr`? It seems like the best place, and you wouldn’t need to reread the `vhdr`. Since the first chunk is 16KB, it should contain the headers, padding, and the beginning of the firmware altogether.

Reply:
i originally didn't want to do that because we limit vendor header size to 64kB actually, in image.c.

On the other hand with vendor header larger thank `(IMAGE_INIT_CHUNK_SIZE- IMAGE_HEADER_SIZE)` this code wont work anyway so it seems to me that we should limit vendor header size to match this.

The 16kB alignment should be enough for any padding due to vector table requirements. 

So i moved the check here https://github.com/trezor/trezor-firmware/pull/4090/commits/65a7ac7e040208eb809d5923b28b855b11ada47d

And i will look into the vendor header size limit in successive PR.